proposals/0004-FaultTolerance.md

```diff
@@ -0,0 +1,111 @@
+# Fault Tolerance
+
+* Proposal: [MP-0004](0004-FaultTolerance.md)
+* Authors: [Emily Jiang](https://github.com/Emily-Jiang), [Jonathan Halterman](https://github.com/jhalterman/)
+* Status: **Awaiting review**
+
+*During the review process, add the following fields as needed:*
+
+* Decision Notes: [Discussion thread topic covering the  Rationale](https://groups.google.com/forum/#!topic/microprofile/ezFC1TLGozU), [Discussion thread topic with additional Commentary](https://groups.google.com/forum/#!forum/microprofile)
+
+## Introduction
+
+It is increasingly important to build fault tolerant micro services. Fault tolerance is about leveraging different strategies to guide the execution and result of some logic. Retry policies, bulkheads, and circuit breakers are popular concepts in this area. They dictate whether and when executions should take place, and fallbacks offer an alternative result when an execution does not complete successfully. 
+
+As mentioned above, the Fault Tolerance proposal is to focus the aspects: TimeOut, RetryPolicy, Fallback, bulkhead and circuit breaker.
+
+* TimeOut: Define a duration for timeout
+* RetryPolicy: Define a criteria on when to retry 
+* Fallback: provide an alternative solution for a failed execution.
+* Bulkhead: isolate failures in part of the system while the rest part of the system can still function.
+* CircuitBreaker: offer a way of fail fast by automatically failing execution to prevent the system overloading and indefinite wait or timeout by the clients.
+
+The main design is to separate execution logic from execution. The execution can be configured with fault tolerance policies, such as RetryPolicy, fallback, Bulkhead and CircuitBreaker. 
+
+[Hystrix](https://github.com/Netflix/Hystrix) and [Failsafe](https://github.com/jhalterman/failsafe) are two popular libraries for handling failures. This proposal is to define a standard API and approach for applications to follow in order to achieve the fault tolerance.
+
+The requirements are as follows:
+
+* Loose coupling: Execution logic should not know anything about the execution status or fault tolerance. 
+* Failure handling strategy should be configured when the execution takes place.
+* Support for synchronous and asynchronous execution
```

Comment:
@viktorklang That's the idea. It isn't a perfect comparison, but conceptually I think of the proposed execution concept as being similar to Scala's Try monad, giving the ability to react to a failure in some common way without having to write a bunch of boilerplate.